src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +194,193 @@ end
+        # must call NamedTuple as function
+        # (e.g. if T contains abstract types, typeof(values) must be a subtype
+        # of but not equal to T because its parameters will all be concrete
+        # types)
+        return new{N,T}(parent, idx, NamedTuple{N,T}(values))
+    end
+end
+
+GroupKey{N}(parent, idx, values::T) where {N,T<:Tuple} = GroupKey{N,T}(parent, idx, values)
+GroupKey(parent, idx, values::NamedTuple{N, T}) where {N,T} = GroupKey{N,T}(parent, idx, values)
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey(")
+    show(io, getfield(k, :values))
+    print(io, ")")
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey{N,T}) where {N,T} = length(T)
+Base.names(::GroupKey{N}) where N = N
+Base.keys(key::GroupKey) = names(key)
+Base.values(key::GroupKey) = values(getfield(key, :parent))
+
+Base.getindex(key::GroupKey, i::Union{Integer,Symbol}) = getfield(key, :values)[i]
+Base.getproperty(key::GroupKey, p::Symbol) = getproperty(getfield(key, :values), p)
+
+Base.:(==)(k1::GroupKey, k2::GroupKey) = parent(k1) === parent(k2) && getfield(k1, :idx) == getfield(k2, :idx)
+
+Base.NamedTuple(key::GroupKey) = getfield(key, :values)
+Base.Tuple(key::GroupKey) = Tuple(getfield(key, :values))
+
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+# Iterator over values of grouping columns for each group
+function _groupvalues(gd::GroupedDataFrame)
+    T = Tuple{_grouptypes(gd)...}
+    return (T(gd.parent[gd.idx[s], gd.cols]) for s in gd.starts)
+end
+
+
+"""
+    keys(gd::GroupedDataFrame)
+
+Get a set of keys for each group of the `GroupedDataFrame`. The keys can be used
+to retrieve the values of the grouping columns for each group and also to index
+the `GroupedDataFrame` effeciently.
+
+### Returns
+
+Vector of [`GroupKey`](@ref)s.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[2]),
+                      b = repeat([2, 1], outer=[3]),
+                      c = 1:6);
+
+julia> gd = groupby(df, :a)
+GroupedDataFrame with 3 groups based on key: a
+First Group (2 rows): a = :foo
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 1     │ 4     │
+⋮
+Last Group (2 rows): a = :baz
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 2     │ 3     │
+│ 2   │ baz    │ 1     │ 6     │
+
+julia> keys(gd)
+3-element Array{DataFrames.GroupKey{(:a,),Tuple{Symbol}},1}:
+ GroupKey((a = :foo,))
+ GroupKey((a = :bar,))
+ GroupKey((a = :baz,))
+```
+"""
+function Base.keys(gd::GroupedDataFrame)
+    N = Tuple(_names(gd)[i] for i in gd.cols)
+    V = Tuple{_grouptypes(gd)...}
+    return [GroupKey{N, V}(gd, i, values) for (i, values) in enumerate(_groupvalues(gd))]
```

Comment:
Better return a generator, allocating a full array is going to be quite costly. Also, why generate `values` upfront? We can defer this until `getproperty` is called on the `GroupKey` (which may never happen).

Reply:
I have made value access lazy, do you think it is still necessary to return a generator here?

Reply:
Yes, it makes even more sense, so that everything is lazy.